File: overlay-api/src/scripts/extract_datas.py
import json
import sys
from openpyxl import load_workbook

if len(sys.argv) < 3:
    print("Usage: python script.py <chemin_excel> <chemin_json>")
    sys.exit(1)

chemin_excel = sys.argv[1]
chemin_json = sys.argv[2]

# Load the workbook with data_only=True to read calculated values
wb = load_workbook(chemin_excel, data_only=True)
ws = wb.active

# Column indices
fields_index = {
    "club": 1,
    "sex": 2,
    "category_age": 4,
    "last_name": 5,
    "first_name": 6,
    "weight_cat": 8,
    "squat": [11, 12, 13],
    "bench": [14, 15, 16],
    "deadlift": [17, 18, 19],
    "total": 20
}

def get_attempt_status(cell):
    fill = cell.fill
    strike = cell.font.strike

    color = fill.start_color.index if fill.start_color else None

    if color in ('0000000B', 11):  # vert / valide
        return "valid"
    elif color in ('0000001F', 31) and strike:  # violet + barré = échec
        return "invalid"
    elif color in ('00000000', None):
        return "pending"
    else:
        return "pending"

athletes = []

for row in ws.iter_rows(min_row=8):
    if not row[fields_index["last_name"]].value or not row[fields_index["first_name"]].value:
        continue

    athlete = {
        "club": row[fields_index["club"]].value,
        "sex": row[fields_index["sex"]].value,
        "category_age": row[fields_index["category_age"]].value,
        "last_name": row[fields_index["last_name"]].value,
        "first_name": row[fields_index["first_name"]].value,
        "weight_category": row[fields_index["weight_cat"]].value,
        "attempts": {
            "squat": [],
            "bench_press": [],
            "deadlift": []
        },
        "total": row[fields_index["total"]].value
    }

    for col in fields_index["squat"]:
        cell = row[col]
        athlete["attempts"]["squat"].append({
            "weight": cell.value,
            "status": get_attempt_status(cell)
        })

    for col in fields_index["bench"]:
        cell = row[col]
        athlete["attempts"]["bench_press"].append({
            "weight": cell.value,
            "status": get_attempt_status(cell)
        })

    for col in fields_index["deadlift"]:
        cell = row[col]
        athlete["attempts"]["deadlift"].append({
            "weight": cell.value,
            "status": get_attempt_status(cell)
        })

    athletes.append(athlete)

# Export the results to JSON
with open(chemin_json, "w", encoding="utf-8") as f:
    json.dump(athletes, f, indent=2, ensure_ascii=False)

File: overlay-api/src/scripts/extract_datas.js
const ExcelJS = require('exceljs');
const fs = require('fs');

/**
 * Détermine le statut d'un essai (valid, invalid, pending)
 */
function getAttemptStatus(cell) {
    const fill = cell.fill;
    const font = cell.font;
    let color = null;

    if (fill && fill.type === 'pattern' && fill.fgColor) {
        // ExcelJS peut retourner l'ARGB ou l'index
        color = fill.fgColor.argb || fill.fgColor.indexed;
    }

    const isStrikethrough = font && font.strike;

    // Vert (Valide) : Index 11 ou ARGB vert standard
    if (color === 'FF00FF00' || color === '0000000B' || color === 11) {
        return "valid";
    } 
    // Violet + Barré (Échec) : Index 31 ou ARGB violet
    else if (isStrikethrough && (color === 'FF800080' || color === '0000001F' || color === 31)) {
        return "invalid";
    } 
    
    return "pending";
}

/**
 * Logique principale de transformation
 */
exports.extractExcelData = async (excelPath, jsonPath) => {
    const workbook = new ExcelJS.Workbook();
    await workbook.xlsx.readFile(excelPath);
    const ws = workbook.getWorksheet(1);

    // On ajuste les indices : si votre script python utilisait row[1], 
    // en ExcelJS getCell(1) correspond à la colonne A. 
    // Si row[1] en python était la colonne B, alors utilisez l'index python + 1.
    const fields = {
        club: 2, sex: 3, catAge: 5, lastName: 6, firstName: 7, 
        weightCat: 9, squat: [12, 13, 14], bench: [15, 16, 17], 
        deadlift: [18, 19, 20], total: 21
    };

    const athletes = [];

    ws.eachRow((row, rowNumber) => {
        if (rowNumber < 8) return;

        const lastName = row.getCell(fields.lastName).value;
        const firstName = row.getCell(fields.firstName).value;
        if (!lastName || !firstName) return;

        const getVal = (cellIndex) => {
            const cell = row.getCell(cellIndex);
            // Gère les cellules avec formules
            return cell.value && typeof cell.value === 'object' ? cell.value.result : cell.value;
        };

        const athlete = {
            club: getVal(fields.club),
            sex: getVal(fields.sex),
            category_age: getVal(fields.catAge),
            last_name: lastName,
            first_name: firstName,
            weight_category: getVal(fields.weightCat),
            attempts: {
                squat: fields.squat.map(col => ({ weight: getVal(col), status: getAttemptStatus(row.getCell(col)) })),
                bench_press: fields.bench.map(col => ({ weight: getVal(col), status: getAttemptStatus(row.getCell(col)) })),
                deadlift: fields.deadlift.map(col => ({ weight: getVal(col), status: getAttemptStatus(row.getCell(col)) }))
            },
            total: getVal(fields.total)
        };
        athletes.push(athlete);
    });

    // Sauvegarde physique du JSON
    fs.writeFileSync(jsonPath, JSON.stringify(athletes, null, 2), 'utf8');
    
    return athletes;
};